File: searching-algorithms/binary-search/binary-search-recursive.c
#include<stdio.h>
#include<stdlib.h>
int checkSorted(int *arr, int n)
{
    int count=0;
    for(int i=0;i<n;i++){
        if(arr[i]>arr[i+1])
        {
            count ++;
        }
    }
    return count;
}
int binarySearch(int array[], int x, int low, int high) {
  if (high >= low) {
    int mid = low + (high - low) / 2;

    // If found at mid, then return it
    if (array[mid] == x)
      return mid;

    // Search the left half
    if (array[mid] > x)
      return binarySearch(array, x, low, mid - 1);

    // Search the right half
    return binarySearch(array, x, mid + 1, high);
  }

  return -1;
}
void merge(int a[],int i1,int j1,int i2,int j2)
{
    int temp[50];    //array used for merging
    int i,j,k;
    i=i1;    //beginning of the first list
    j=i2;    //beginning of the second list
    k=0;
    while(i<=j1 && j<=j2)    //while elements in both lists
    {
        if(a[i]<a[j])
            temp[k++]=a[i++];
        else
            temp[k++]=a[j++];
    }
    while(i<=j1)    //copy remaining elements of the first list
        temp[k++]=a[i++];
    while(j<=j2)    //copy remaining elements of the second list
        temp[k++]=a[j++];
    //Transfer elements from temp[] back to a[]
    for(i=i1,j=0;i<=j2;i++,j++)
        a[i]=temp[j];
}


void mergesort(int a[],int i,int j)
{
    int mid;
    if(i<j)
    {
        mid=(i+j)/2;
        mergesort(a,i,mid);        //left recursion
        mergesort(a,mid+1,j);    //right recursion
        merge(a,i,mid,mid+1,j);    //merging of two sorted sub-arrays
    }
}
void printArray(int array[], int n) 
{ 
    int i; 
    for (i = 0; i < n; i++) 
        printf("%d ", array[i]); 
    printf("\n"); 
}
int main()
{
    system("cls");
    int arr[100], n,x;
    // ask user for number of elements to be sorted
    printf("Enter the number of elements: ");
    scanf("%d", &n);
    // input elements if the array
    for(int i=0;i<n;i++)
    {
        printf("Enter element no. %d: ", i+1);
        scanf("%d", &arr[i]);
    }
    printf("\nEnter the element to be found: ");
    scanf("%d", &x);
    int sorted= checkSorted(arr,n);
    if(sorted!=0){
        printf("Array before sorting: ");
        printArray(arr,n);
        mergesort(arr,0,n-1);
        printf("Array after sorting: ");
        printArray(arr,n);
        printf("\n\n");
        int result=binarySearch(arr,x,0,n-1);
        printf("\n\n");
        if(result==-1){
            printf("NOT FOUND!!");
        }
        else{
            printf("Element found at (in the sorted list): %d",result+1);
        }
    }
    else{
        printf("Array: \n");
        printArray(arr,n);
        int result=binarySearch(arr,x,0,n-1);
        if(result==-1){
            printf("NOT FOUND!!");
        }
        else{
            printf("Element found at: %d",result+1);
        }
    }
    return 0;
}

File: searching-algorithms/binary-search/binary-search-recursive.py
def checkSorted(arr,n):
    flag=0
    if(arr==sorted(arr)):
        flag =1
    return flag


def binarySearch(array, x, low, high):

    if high >= low:

        mid = low + (high - low)//2

        # If found at mid, then return it
        if array[mid] == x:
            return mid

        # Search the left half
        elif array[mid] > x:
            return binarySearch(array, x, low, mid-1)

        # Search the right half
        else:
            return binarySearch(array, x, mid + 1, high)

    else:
        return -1

def merge_sort(inp_arr):
    size = len(inp_arr)
    if size > 1:
        middle = size // 2
        left_arr = inp_arr[:middle]
        right_arr = inp_arr[middle:]
 
        merge_sort(left_arr)
        merge_sort(right_arr)
 
        p = 0
        q = 0
        r = 0
 
        left_size = len(left_arr)
        right_size = len(right_arr)
        while p < left_size and q < right_size:
            if left_arr[p] < right_arr[q]:
              inp_arr[r] = left_arr[p]
              p += 1
            else:
                inp_arr[r] = right_arr[q]
                q += 1
             
            r += 1
 
        
        while p < left_size:
            inp_arr[r] = left_arr[p]
            p += 1
            r += 1
 
        while q < right_size:
            inp_arr[r]=right_arr[q]
            q += 1
            r += 1
def printArr(arr):
    for element in arr:
        print(element,end=" ")
    print("")

n = int(input("Enter the size of the list "))
print("\n")
arr = list(int(num) for num in input("Enter elements ").strip().split())[:n]
search=int(input("Enter the value to be searched: "))
sorted= checkSorted(arr,n)
if (not sorted):
    print("Unsorted List: ")
    printArr(arr)
    merge_sort(arr)
    print("\n")
    printArr(arr)
    result = binarySearch(arr, search, 0, len(arr)-1)

    if result != -1:
        print("Element is present at index " + str(result+1))
    else:
        print("Not found")
else:
    result = binarySearch(arr, search, 0, len(arr)-1)

    if result != -1:
        print("Element is present at index " + str(result+1))
    else:
        print("Not found")

